modules/4337/certora/specs/SignatureLengthCheck.spec

```diff
@@ -0,0 +1,18 @@
+using Account as safeContract;
+
+methods {
+    function checkSignaturesLength(bytes, uint256) external returns(bool) envfree;
+    function combineBytes(bytes, bytes) external returns(bytes) envfree;
+
+    // Safe Contract functions
+    function safeContract.canonicalSignatureHash(bytes, uint256) external returns(bytes32) envfree;
+}
+
+// This rule verifies that if excess data is added to the signature, though it could pass in the safe contract's `checkSignatures(...)`,
+// it will be caught within the `_checkSignaturesLength(...)` call, as the dynamic length is checked.
+rule canonicalHashBasedLengthCheck(bytes signatures, bytes griefedSignatures, uint256 threshold) {
+    require safeContract.canonicalSignatureHash(signatures, threshold) == safeContract.canonicalSignatureHash(griefedSignatures, threshold);
+    require signatures.length < griefedSignatures.length;
+
+    assert checkSignaturesLength(signatures, threshold) => !checkSignaturesLength(griefedSignatures, threshold);
```

Comment:
I'm surprised this doesn't find a counter example, as I think I can come up with one.

Reply:
We `require` in the canonical hash function for the signature length to be > 0

https://github.com/safe-global/safe-modules/blob/f72107828db4c06d3a81ee1563dca721a54e845a/modules/4337/certora/harnesses/Account.sol#L92

So, the above example is not valid.

Reply:
Why? Safe contract signatures don't require this:

https://github.com/safe-global/safe-smart-account/blob/8f80a8372d193be121dcdb52e869a258824e5c0f/contracts/Safe.sol#L231

The exact same counter example can also exist for signatures of any length - the "trick" is that two signers have identical signatures:

```
"0000000000000000000000000000000000000000000000000000000000000001" + (65*2).toString(16).padStart(64, "0") + "00" +
"0000000000000000000000000000000000000000000000000000000000000002" + (65*2).toString(16).padStart(64, "0") + "00" +
"0000000000000000000000000000000000000000000000000000000000000001" + "ff"
```

and

```
"0000000000000000000000000000000000000000000000000000000000000001" + (65*2).toString(16).padStart(64, "0") + "00" +
"0000000000000000000000000000000000000000000000000000000000000002" + (65*2 + 32).toString(16).padStart(64, "0") + "00" +
"0000000000000000000000000000000000000000000000000000000000000001" + "ff" +
"0000000000000000000000000000000000000000000000000000000000000001" + "ff"
```

Reply:
Updated, thanks for the pair programming session @nlordell!